### Classification Metrics


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_true = [1, 0, 1, 0, 1, 0, 1, 1]
y_pred = [1, 0, 0, 0, 0, 1, 0, 1]

accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred)
recallScore = recall_score(y_true, y_pred)
f1Score = f1_score(y_true, y_pred)
confusionMetricsTable = confusion_matrix(y_true, y_pred)

print("accuracy:", accuracy)                                   
print("precision:", precision)
print("recallScore:", recallScore)
print("f1Score:", f1Score)
print("confusionMetricsTable:", confusionMetricsTable)      

accuracy: 0.5
precision: 0.6666666666666666
recallScore: 0.4
f1Score: 0.5
confusionMetricsTable: [[2 1]
 [3 2]]


Confuction Metrics Table
|               | **Predicted: 1**          | **Predicted: 0**             |
| ------------- | ------------------------- | ---------------------------- |
| **Actual: 1** | **TP = 2** (indices 0, 7) | **FN = 3** (indices 2, 4, 6) |
| **Actual: 0** | **FP = 1** (index 5)      | **TN = 2** (indices 1, 3)    |

Metrics Table: 
| Metric        | Formula                         | Calculation             | **Result** |
| ------------- | ------------------------------- | ----------------------- | ---------- |
| **Accuracy**  | (TP + TN) / (TP + TN + FP + FN) | (2 + 2) / 8             | **0.5**    |
| **Precision** | TP / (TP + FP)                  | 2 / (2 + 1)             | **0.67**   |
| **Recall**    | TP / (TP + FN)                  | 2 / (2 + 3)             | **0.4**    |
| **F1-Score**  | 2 × (P × R) / (P + R)           | 2×(0.67×0.4)/(0.67+0.4) | **0.5**    |


- Accuracy:
    - Meaning: % of correct predictions
    - Use Case: Balanced datasets
    - Ideal Value: 1.0 (100%)

- Precision:
    - Meaning: Of all predicted positives, how many are actually positive
    - Use Case: When False Positives matter (e.g. spam filter)
    - Ideal Value: 1.0

- Recall (Sensitivity):
    - Meaning: Of all actual positives, how many did we catch
    - Use Case: When False Negatives matter (e.g. disease detection)
    - Ideal Value: 1.0

- F1-Score:
    - Meaning: Harmonic mean of Precision & Recall
    - Use Case: Unbalanced datasets
    - Ideal Value: 1.0

- Confusion Matrix:
    - Meaning: Table showing TP, TN, FP, FN counts
    - Use Case: To see class-wise errors
